File: typescript/Y2022/Day4.ts
import { Solution } from '../run';

type Range = [number, number];

const parseRow = (row: string): [Range, Range] => {
  return row.split(',').map((x) => x.split('-').map((y) => +y)) as [
    Range,
    Range,
  ];
};

const contained = (a: Range, b: Range): boolean => a[0] <= b[0] && a[1] >= b[1];

const part1 = (input: string) => {
  return input
    .trim()
    .split('\n')
    .map(parseRow)
    .reduce((acc, [a, b]) => acc + +(contained(a, b) || contained(b, a)), 0);
};

const intersects = (a: Range, b: Range): boolean =>
  a[1] >= b[0] && a[0] <= b[1];

const part2 = (input: string) => {
  return input
    .trim()
    .split('\n')
    .map(parseRow)
    .reduce((acc, [a, b]) => acc + +intersects(a, b), 0);
};

const solution: Solution<number, number> = {
  part1,
  part2,
  answers: [526, 886],
};

export default solution;

File: advent_of_rust/src/y2022/day_4.rs
use crate::Solution;

pub fn get_solution() -> Solution<usize, usize> {
    Solution {
        date: (2022, 4),
        part_1: Box::new(part_1),
        part_2: Box::new(part_2),
        answer: (526, 886),
    }
}
type Range = (i32, i32);

fn parse_row(row: &str) -> anyhow::Result<(Range, Range)> {
    let mut pairs = row.split(',');
    let mut a = pairs.next().unwrap().split('-');
    let mut b = pairs.next().unwrap().split('-');

    Ok((
        (a.next().unwrap().parse()?, a.next().unwrap().parse()?),
        (b.next().unwrap().parse()?, b.next().unwrap().parse()?),
    ))
}

fn contained(a: Range, b: Range) -> bool {
    a.0 <= b.0 && a.1 >= b.1
}

fn part_1(input: &str) -> anyhow::Result<usize> {
    let mut sum = 0;

    for line in input.trim().lines() {
        let (a, b) = parse_row(line)?;

        if contained(a, b) || contained(b, a) {
            sum += 1;
        }
    }

    Ok(sum)
}

fn intersects(a: Range, b: Range) -> bool {
    b.0 <= a.1 && b.1 >= a.0
}

fn part_2(input: &str) -> anyhow::Result<usize> {
    let mut sum = 0;

    for line in input.trim().lines() {
        let (a, b) = parse_row(line)?;

        if intersects(a, b) {
            sum += 1;
        }
    }

    Ok(sum)
}
